### PR TITLE
Detect CTRL/SHIFT/ALT + add control dialg + add end-of-race ranking screen + fixes

### LINKED ISSUE #86: 60% of functions implemented

### LINKED ISSUE #104: Fog effect is drawn with an offset when using a smaller screen
The offset is 0 when using full screen.
Also, the black-white change in the background changes when switching screen size.

How to reproduce:
- start dethrace with demo version
- increase/decrease screen with the `-`/`+` button on the left of the backspace key

These screenshots are made with the demo version.

Screenshot from 2022-03-11 16-56-25
Screenshot from 2022-03-11 16-56-28
Screenshot from 2022-03-11 16-56-32

### LINKED ISSUE #111: Switching to cockpit crashes the game
When in-race, pressing <kbd>C</kbd> to enter the cockpit crashes the game.
This is caused by the mirror logic not being implemented yet.

Workaround:

disable the mirror

### PR DESCRIPTION
**Pull request by madebr**
This PR:
- detects CTRL/SHIFT/ALT
- implements QWE for looking around in the cockpit
- starts hooking up some functions for edit mode
- adds the end-of-race screen that calculates the new rank + credits
- add the displays for the demo's that block you from loading/saving/multiplayer
- implements the controls options screen
- fixes #111
- fixes #104 
- fixes #86 (currently at 60%)
- fixes fully displaying the fancy head-ups (they weren't shown when entering the display)
- fixes various AddressSanitizer errors: we were reading and writing out-of-bounds pixels. To fix that, I started implementing and using the pixelmap dispatch table.

I only tested the `mrcursorscool` cheat, as that one is useful to quickly finish a race.

Screenshot from 2022-03-29 20-30-42
Screenshot from 2022-03-29 20-28-28
Screenshot from 2022-03-22 21-22-34

### CODE REVIEW

File: src/DETHRACE/common/displays.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ extern int gTimer_headup;
 extern int gTime_awarded_headup;
 extern int gPed_kill_count_headup;
 extern int gDim_amount;
-extern br_pixelmap* gHeadup_images[31];
+extern br_pixelmap* gHeadup_images[32];
```
Comment (dethrace-labs) [inline src/DETHRACE/common/displays.h:29]: How did you determine this should be 32? `DisposeHeadupImages` is clearly going from index 0 to 30?

Comment (dethrace-labs) [reply src/DETHRACE/common/displays.h:29]: at the end of the array, at 0x128CA4 I have `gNet_cash_headup`

Comment (madebr) [reply src/DETHRACE/common/displays.h:29]: See `LoadHeadupImages` -> `gHeadup_images`.
It's to accommodate the demo timeout fancy head up.

Comment (madebr) [reply src/DETHRACE/common/displays.h:29]: This is demo-only behavior.

File: src/DETHRACE/common/mainmenu.c [hunk 2]
```diff
@@ -30,6 +30,15 @@ char* gPixels_copy__mainmenu; // suffix added to avoid duplicate symbol
 int MainMenuDone1(int pCurrent_choice, int pCurrent_mode, int pGo_ahead, int pEscaped, int pTimed_out) {
     LOG_TRACE("(%d, %d, %d, %d, %d)", pCurrent_choice, pCurrent_mode, pGo_ahead, pEscaped, pTimed_out);
 
+    if (pCurrent_mode == 0) {
```
Comment (madebr) [inline src/DETHRACE/common/mainmenu.c:33]: This looks incorrect.
This is the code for the demo.

File: src/DETHRACE/common/racesumm.c [hunk 3]
```diff
@@ -86,46 +226,207 @@ void DrawSummary(int pCurrent_choice, int pCurrent_mode) {
...
     tS3_sound_tag sound_tag;
     LOG_TRACE("(%p, %p)", pCurrent_choice, pCurrent_mode);
-    NOT_IMPLEMENTED();
+
+    if (gTemp_credits > 0) {
+        return 0;
+    }
+    S3StopOutletSound(gIndexed_outlets[0]);
+    RemoveTransientBitmaps(1);
+    for (i = 0; i < 7; i++) {
```
Comment (dethrace-labs) [inline src/DETHRACE/common/racesumm.c:338]: `i` isn't used anywhere in the loop?

Comment (madebr) [reply src/DETHRACE/common/racesumm.c:338]: Nope, not used.
I think this loop is used to smoothly stop the music.

File: src/DETHRACE/common/world.c [hunk 4]
```diff
@@ -857,7 +857,7 @@ tFunkotronic_spec* AddNewFunkotronic() {
     }
     gFunkotronics_array = new_array;
     for (i = 0; i < 16; i++) {
-        gFunkotronics_array[gFunkotronics_array_size - i].owner = -999;
+        gFunkotronics_array[gFunkotronics_array_size - 16 + i].owner = -999;
```
Comment (dethrace-labs) [inline src/DETHRACE/common/world.c:860]: nice catch

Comment (madebr) [reply src/DETHRACE/common/world.c:860]: Thank AddressSanitizer, he "reminded" me :D

**Review by dethrace-labs — APPROVED** [review]

**Pull request merged**